src/ApiHandler.ts

```diff
@@ -345,13 +352,101 @@ export default class ApiHandler {
 		}
 	}
 
+	/**
+	 * Fetch and derive generalized staking information at a specific block.
+	 *
+	 * @param hash BlockHash to make call at.
+	 */
+	async fetchStakingInfo(hash: BlockHash): Promise<StakingInfo> {
+		const api = await this.ensureMeta(hash);
+
+		const [
+			validatorCount,
+			forceEra,
+			eraElectionStatus,
+			validators,
+			{ number },
+		] = await Promise.all([
+			await api.query.staking.validatorCount.at(hash),
+			await api.query.staking.forceEra.at(hash),
+			await api.query.staking.eraElectionStatus.at(hash),
+			await api.query.session.validators.at(hash),
+			await api.rpc.chain.getHeader(hash),
+		]);
+
+		const queuedElected = await this.tryFetchQueueElected(api, hash);
+
+		const {
+			eraLength,
+			eraProgress,
+			sessionLength,
+			sessionProgress,
+			activeEra,
+		} = await this.deriveSessionAndEraProgress(api, hash);
+
+		const unappliedSlashesAtActiveEra = activeEra
+			? await api.query.staking.unappliedSlashes.at(hash, activeEra)
+			: null;
+
+		const currentBlockNumber = number.toBn();
+
+		const nextSession = sessionLength
+			.sub(sessionProgress)
+			.add(currentBlockNumber);
+
+		// when ForceNone we are in PoA
+		const nextEra = forceEra.isForceNone
+			? null // If forceEra.isForceNone there is no next era (null)
+			: forceEra.isForceAlways
+			? nextSession // If forceEra.isForceAlways, there is a new era every session
+			: eraLength.sub(eraProgress).add(currentBlockNumber); // Otherwise the nextEra is at the end of this era
+
+		const electionLookAhead = await this.deriveElectionLookAhead(api, hash);
+
+		const toggle =
+			forceEra.isForceNone || electionLookAhead.lte(new BN(0))
+				? null // If forceEra.isForceNone there is no next era and thus no upcoming elections
+				: eraElectionStatus.isClose
+				? nextEra?.sub(electionLookAhead) // If the election is closed than it has not happened yet
+				: nextEra; // Otherwise the election is open and it will close at the beginning of the next era
+
+		return {
+			at: {
+				hash: hash.toJSON(),
+				height: currentBlockNumber.toString(10),
+			},
+			validatorCount: forceEra.isForceNone
+				? null
+				: validatorCount.toString(10),
+			activeEra: activeEra.toString(10) ?? null,
```

Comment:
what does ?? do?

Reply:
`(cond) ? true : false`

it reminds me C :-)

Reply:
If the operator on the left is undefined or null, return the operator on the right. I use it often because I want to make sure our API is consistent in using null and we don't have an undefined sneak in.